allennlp/interpret/attack/hotflip.py

```diff
@@ -0,0 +1,184 @@
+from typing import Dict, List, Set
+import numpy
+import torch
+from allennlp.interpret.attack import Attacker
+from allennlp.common.util import JsonDict, sanitize 
+from allennlp.data import Instance
+from allennlp.predictors.predictor import Predictor
+from allennlp.data.fields.field import DataArray, Field
+from allennlp.data.fields import IndexField
+from allennlp.modules.text_field_embedders.text_field_embedder import TextFieldEmbedder
+from allennlp.modules.text_field_embedders import BasicTextFieldEmbedder
+from allennlp.modules.token_embedders import Embedding
+from allennlp.data.tokenizers import Token
+from collections import defaultdict
+
+@Attacker.register('hotflip')
+class Hotflip(Attacker):
+    def __init__(self, predictor):
+        """
+        Runs the HotFlip style attack at the word-level https://arxiv.org/abs/1712.06751.
+        We use the first-order taylor approximation described in https://arxiv.org/abs/1903.06620,
+        in the function first_order_taylor().
+        """
+        super().__init__(predictor)                                          
+        self.vocab = self.predictor._model.vocab
+        """ For HotFlip, we need a word embedding matrix to search over. The below is neccessary for
+        models such as ELMo, character-level models, or for models that use a projection layer
+        after their word embeddings. 
+        We run all of the words from the dataset reader through the TextFieldEmbedder, and save the final
+        output embedding. We then group all of those output embeddings into an "embedding matrix".
+        """
+        # Gets all tokens in the vocab and their corresponding IDs
+        all_tokens = list(self.vocab._token_to_index["tokens"].keys())            
+        all_inputs = {"tokens":torch.LongTensor([x for x in self.vocab._index_to_token["tokens"].keys()]).unsqueeze(0)}          
+        # handle when a model uses character-level inputs, e.g., ELMo or a CharCNN
+        if "token_characters" in self.predictor._dataset_reader._token_indexers:            
+            all_tokens_tokenized = self.predictor._dataset_reader._token_indexers["token_characters"]._character_tokenizer.batch_tokenize(all_tokens)                        
```

Comment:
You can make this a whole lot simpler, using `tokens_to_indices` like you do for elmo:

```python
indexer = self.predictor._dataset_reader._token_indexers["token_characters"]
tokens = [Token(x) for x in all_tokens]
max_token_length = max(len(x) for x in all_tokens)
indexed_tokens = indexer.tokens_to_indices(tokens, self.vocab, "token_characters")
padded_tokens = indexer.pad_token_sequence(indexed_tokens,
                                           desired_num_tokens={"token_characters": len(tokens)},
                                           padding_lengths={"num_token_characters": max_token_length})
all_inputs.update(padded_tokens)
```